Given an integer `n`, return *`true` if it is a power of two. Otherwise, return `false`*.

An integer `n` is a power of two, if there exists an integer `x` such that `n == 2x`.

**Example 1:**
Input: n = 1 

Output: true

**Example 2:**
Input: n = 16 

Output: true

**Example 3:**
Input: n = 3 

Output: false

In [1]:
def checkPowerOf2(n:int())-> int():
  if n==2:
    return 2
  if n < 2:
    return -1
  return checkPowerOf2(n/2)

In [2]:
if checkPowerOf2(25) == 2:
  print(True)
else:
  print(False)

False


<aside>
💡 **Question 2**

Given a number n, find the sum of the first natural numbers.

**Example 1:**

Input: n = 3 

Output: 6

**Example 2:**

Input  : 5 

Output : 15

</aside>

In [6]:
def sum(n:int())->int(): 
  return n+sum(n-1) if n >0 else 0
sum(15)

120

<aside>
💡 **Question 3**

****Given a positive integer, N. Find the factorial of N. 

**Example 1:**

Input: N = 5 

Output: 120

**Example 2:**

Input: N = 4

Output: 24

</aside>

In [8]:
def sum(n:int())->int(): 
  return n*sum(n-1) if n >0 else 1
sum(5)

120

<aside>
💡 **Question 4**

Given a number N and a power P, the task is to find the exponent of this number raised to the given power, i.e. N^P.

**Example 1 :** 

Input: N = 5, P = 2

Output: 25

**Example 2 :**
Input: N = 2, P = 5

Output: 32

</aside>

In [10]:
def sum(n:int(), p:int())->int(): 
  return n*sum(n, p-1) if p>0 else 1
sum(5, 2)

25

<aside>
💡 **Question 5**

Given an array of integers **arr**, the task is to find maximum element of that array using recursion.

**Example 1:**

Input: arr = {1, 4, 3, -5, -4, 8, 6};
Output: 8

**Example 2:**

Input: a9rr = {1, 4, 45, 6, 10, -8};
Output: 45

</aside>

In [21]:
def findMax(arr:list(), curMax: int()) ->int():
  if len(arr) ==1:
    return curMax if curMax> arr[0] else arr[0]
  return curMax if curMax> findMax(arr[1:], curMax) else findMax(arr[1:], curMax)

In [23]:
findMax([1,2,3,5,1,2,3,252,2,356], 3)

356

<aside>
💡 **Question 6**

Given first term (a), common difference (d) and a integer N of the Arithmetic Progression series, the task is to find Nth term of the series.

**Example 1:**

Input : a = 2 d = 1 N = 5
Output : 6
The 5th term of the series is : 6

**Example 2:**

Input : a = 5 d = 2 N = 10
Output : 23
The 10th term of the series is : 23

</aside>

In [30]:
def termOfAp(a: int(), d:int(), n:int()) -> int():
  return a+(d*(n-1))

In [31]:
a = 10
n = 5
d = 3

print(termOfAp(a,d,n))

22


<aside>
💡 **Question 7**

Given a string S, the task is to write a program to print all permutations of a given string.

**Example 1:**

***Input:***

*S = “ABC”*

***Output:***

*“ABC”, “ACB”, “BAC”, “BCA”, “CBA”, “CAB”*

**Example 2:**

***Input:***

*S = “XY”*

***Output:***

*“XY”, “YX”*

</aside>

In [32]:
def permute(s):
  out = []
  if len(s) == 1:
    return s
  else:
    for i,let in enumerate(s):
      for perm in permute(s[:i] + s[i+1:]):
        out += [let + perm]
  return out

In [34]:
permute(['1','2','3','6'])

['1236',
 '1263',
 '1326',
 '1362',
 '1623',
 '1632',
 '2136',
 '2163',
 '2316',
 '2361',
 '2613',
 '2631',
 '3126',
 '3162',
 '3216',
 '3261',
 '3612',
 '3621',
 '6123',
 '6132',
 '6213',
 '6231',
 '6312',
 '6321']

<aside>
💡 **Question 8**

Given an array, find a product of all array elements.

**Example 1:**

Input  : arr[] = {1, 2, 3, 4, 5}
Output : 120
**Example 2:**

Input  : arr[] = {1, 6, 3}
Output : 18

</aside>

In [38]:
def prod(arr:list())->int(): 
  return arr[0] * prod(arr[1:]) if len(arr) >2 else arr[0]*arr[1]
prod([1,3,5,4,5])

300